File: Ex2.go
package main

import (
  "fmt"
)

// Функция для преобразования римского числа в натуральное число
func romanToInt(romanNumeral string) int {
  // Создаем словарь для соответствия римских цифр и их значений
  romanNumerals := map[byte]int{
    'I': 1,
    'V': 5,
    'X': 10,
    'L': 50,
    'C': 100,
    'D': 500,
    'M': 1000,
  }

  result := 0
  for i := 0; i < len(romanNumeral); i++ {
    current := romanNumerals[romanNumeral[i]]
    if i+1 < len(romanNumeral) && current < romanNumerals[romanNumeral[i+1]] {
      result += romanNumerals[romanNumeral[i+1]] - current
      i++ // Пропускаем следующую цифру, так как она уже обработана
    } else {
      result += current
    }
  }

  return result
}

func main() {
  // Получаем ввод от пользователя
  var romanNumeral string
  fmt.Print("Введите римское число: ")
  fmt.Scan(&romanNumeral)

  // Преобразуем римское число в натуральное и выводим результат
  fmt.Println("Результат:", romanToInt(romanNumeral))
}

File: Ex2.rs
use std::collections::HashMap;
use std::io;

// Функция для преобразования римского числа в натуральное число
fn roman_to_int(roman_numeral: &str) -> i32 {
    // Создаем HashMap для соответствия римских цифр и их значений
    let mut roman_numerals = HashMap::new();
    roman_numerals.insert('I', 1);
    roman_numerals.insert('V', 5);
    roman_numerals.insert('X', 10);
    roman_numerals.insert('L', 50);
    roman_numerals.insert('C', 100);
    roman_numerals.insert('D', 500);
    roman_numerals.insert('M', 1000);

    let mut result = 0;
    let mut prev = 0;
    for char in roman_numeral.chars().rev() {
        let current = roman_numerals.get(&char).unwrap();
        if *current < prev {
            result -= current;
        } else {
            result += current;
        }
        prev = *current;
    }
    result
}

fn main() {
    // Получаем ввод от пользователя
    println!("Введите римское число:");
    let mut roman_numeral = String::new();
    io::stdin().read_line(&mut roman_numeral)
        .expect("Ошибка при чтении строки");
    let roman_numeral = roman_numeral.trim().to_uppercase();

    // Преобразуем римское число в натуральное и выводим результат
    let result = roman_to_int(&roman_numeral);
    println!("Результат: {}", result);
}
